doc/miner.md

```diff
@@ -7,6 +7,95 @@ Bitcoin Unlimited is based on the Satoshi codebase, so it is a drop in replaceme
 
 But Bitcoin Unlimited has specific features to facilitate mining.
 
+***getminingcandidate*** and ***submitminingsolution***
+----------------------------------
+*efficient protocol to access block candidates and submit block solutions*
+
+
+Bitcoin Unlimited provides 2 additional mining RPC functions that can be used instead of "getblocktemplate" and "submitblock".  These RPCs do not pass the entire block to mining pools.  Instead, the candidate block header, proposed coinbase transaction, and coinbase merkle proof are passed.  This is the approximately the same data that is passed to hashing hardware via the Stratum protocol, so if you are familiar with Stratum, you are familiar with how this is possible.
+
+A mining pool uses ***getminingcandidate*** to receive the previously described block information and a tracking identifier.  It then may modify or completely replace the coinbase transaction and many block header fields, to create different candidates for hashing hardware.  It then forwards these candidates to the hashing hardware via Stratum.  When a solution is found, the mining pool can submit the solution back to bitcoind via submitminingsolution.
+
+A few of the benefits when using RPC getminingcandidate and RPC submitminingsolution are:
+* Massively reduced bandwidth and latency, especially for large blocks.  This RPC requires log2(blocksize) data. 
+* Faster JSON parsing and creation
+* Concise JSON
+
+### bitcoin-miner
+
+An example CPU-miner program is provided that shows a proof-of-concept use of these functions.
+The source code is located at bitcoin-miner.cpp.  To try it out, run
+```sh
+bitcoin-miner
+```
+
+Of course, given current and foreseeable mining difficulties this program will not find any blocks on mainnet.  However, it will find blocks on testnet or regtest.
+
+### miningtest.py
+
+A python based test of these interfaces is located at qa/rpc-tests/miningtest.py.  This example may be of more use for people accessing these RPCs in higher level languages.
+
+### Function documentation:
+
+#### RPC getminingcandidate
+
+##### Arguments: -none
+##### Returns:
+```
+{
+  # candidate identifier for submitminingsolution (integer):
+  "id": 14,
+  
+  # Hash of the previous block (hex string):
+  "prevhash": "0000316517e048ab283a41df3c0ba125345a5c56ef3f76db901b0ede65e2f0e5",
+  
+  # Coinbase transaction (hex string encoded binary transaction)
+  "coinbase": "...00ffffffff10028122000b2f454233322f414431322ffff..."
+
+  # Block version (integer):
+  "version": 536870912,
+  
+  # Difficulty (hex string):
+  "nBits": "207fffff",
+  
+  # Block time (integer):
+  "time": 1528925409,
+  
+  # Merkle branches for the block (list of hex strings):
+  "merklebranches": [
```

Comment:
If the miner has the merkle root and path to the coinbase, listing all of the branches shouldn't be necessary.

Also, an indication of where the coinbase is located within the merkle tree will make this API more robust against potential changes in the future that modify transaction ordering.  See how we're planning on solving that for ABC: https://docs.google.com/document/d/10YVLdltuppEoizAVoovPTOZbqthbsa2t0MXHhaDo4bI/

Reply:
This is just the path to the coinbase.  It contains the merkle root of the all the branches of the tree (subtrees) that the miner does not need.  I think that creating a list of dictionaries to solve the non-0 coinbase is excessively complicated and will increase parsing time.  Instead think about adding a single integer where each bit tells you to to CAT(A,B) or CAT(B,A).